adapter/statsd/statsd.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statsd
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"text/template"
+	"time"
+
+	"github.com/cactus/go-statsd-client/statsd"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/duration"
+	"github.com/hashicorp/go-multierror"
+
+	"istio.io/mixer/adapter/statsd/config"
+	"istio.io/mixer/pkg/adapter"
+)
+
+const (
+	defaultFlushBytes = 512
+)
+
+type (
+	builder struct {
+		adapter.DefaultBuilder
+	}
+
+	aspect struct {
+		rate      float32
+		client    statsd.Statter
+		templates map[string]*template.Template // metric name -> template
+	}
+)
+
+var (
+	name        = "statsd"
+	desc        = "Pushes statsd metrics"
+	defaultConf = &config.Params{
+		Address:                   "localhost:8125",
+		Prefix:                    "",
+		FlushDuration:             &duration.Duration{Nanos: int32(300 * time.Millisecond)},
+		FlushBytes:                512,
+		SamplingRate:              1.0,
+		MetricNameTemplateStrings: make(map[string]string),
+	}
+)
+
+// Register records the builders exposed by this adapter.
+func Register(r adapter.Registrar) {
+	r.RegisterMetricsBuilder(newBuilder())
+}
+
+func newBuilder() *builder {
+	return &builder{adapter.NewDefaultBuilder(name, desc, defaultConf)}
+}
+
+func (b *builder) ValidateConfig(c adapter.AspectConfig) (ce *adapter.ConfigErrors) {
+	params := c.(*config.Params)
+	flushDuration, err := ptypes.Duration(params.FlushDuration)
+	if err != nil {
+		ce = ce.Append("FlushDuration", err)
+	}
+	if flushDuration < time.Duration(0) {
+		ce = ce.Append("FlushDuration", fmt.Errorf("flush duration must be >= 0"))
+	}
+	if params.FlushBytes < 0 {
+		ce = ce.Append("FlushBytes", fmt.Errorf("flush bytes must be >= 0"))
+	}
+	if params.SamplingRate < 0 {
+		ce = ce.Append("SamplingRate", fmt.Errorf("sampling rate must be >= 0"))
+	}
+	if _, err := net.ResolveUDPAddr("udp", params.Address); err != nil {
```

Comment:
As before, I think ValidateConfig is about format verification and not about runtime verification. In other words, ValidateConfig should be fully deterministic. Checking if an address can resolve is non-deterministic and should therefore be treated as a runtime error instead.